server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorPlumber.java

```diff
@@ -0,0 +1,501 @@
+  private final Appenderator appenderator;
+
+  private volatile boolean shuttingDown = false;
+  private volatile boolean stopped = false;
+  private volatile boolean cleanShutdown = true;
+  private volatile ScheduledExecutorService scheduledExecutor = null;
+
+  private volatile Supplier<Committer> lastCommitterSupplier = null;
+
+  public AppenderatorPlumber(
+      DataSchema schema,
+      RealtimeTuningConfig config,
+      FireDepartmentMetrics metrics,
+      DataSegmentAnnouncer segmentAnnouncer,
+      SegmentPublisher segmentPublisher,
+      SegmentHandoffNotifier handoffNotifier,
+      Appenderator appenderator
+  )
+  {
+    this.schema = schema;
+    this.config = config;
+    this.rejectionPolicy = config.getRejectionPolicyFactory().create(config.getWindowPeriod());
+    this.metrics = metrics;
+    this.segmentAnnouncer = segmentAnnouncer;
+    this.segmentPublisher = segmentPublisher;
+    this.handoffNotifier = handoffNotifier;
+    this.appenderator = appenderator;
+
+    log.info("Creating plumber using rejectionPolicy[%s]", getRejectionPolicy());
+  }
+
+  public Map<Long, SegmentIdentifier> getSegmentsView() {
+    return ImmutableMap.copyOf(segments);
+  }
+
+  public DataSchema getSchema()
+  {
+    return schema;
+  }
+
+  public RealtimeTuningConfig getConfig()
+  {
+    return config;
+  }
+
+  public RejectionPolicy getRejectionPolicy()
+  {
+    return rejectionPolicy;
+  }
+
+  @Override
+  public Object startJob()
+  {
+    handoffNotifier.start();
+    Object retVal = appenderator.startJob();
+    initializeExecutors();
+    startPersistThread();
+    // Push pending sinks bootstrapped from previous run
+    mergeAndPush();
+    return retVal;
+  }
+
+  @Override
+  public int add(InputRow row, Supplier<Committer> committerSupplier) throws IndexSizeExceededException
+  {
+    final SegmentIdentifier identifier = getSegmentIdentifier(row.getTimestampFromEpoch());
+    if (identifier == null) {
+      return -1;
+    }
+
+    final int numRows;
+
+    try {
+      numRows = appenderator.add(identifier, row, committerSupplier);
+      lastCommitterSupplier = committerSupplier;
+      return numRows;
+    }
+    catch (SegmentNotWritableException e) {
+      // Segment already started handoff
+      return -1;
+    }
+  }
+
+  @Override
+  public <T> QueryRunner<T> getQueryRunner(final Query<T> query)
+  {
+    return new QueryRunner<T>()
+    {
+      @Override
+      public Sequence<T> run(final Query<T> query, final Map<String, Object> responseContext)
+      {
+        return query.run(appenderator, responseContext);
+      }
+    };
+  }
+
+  @Override
+  public void persist(final Committer committer)
+  {
+    final Stopwatch runExecStopwatch = Stopwatch.createStarted();
+    appenderator.persistAll(committer);
+
+    final long startDelay = runExecStopwatch.elapsed(TimeUnit.MILLISECONDS);
+    metrics.incrementPersistBackPressureMillis(startDelay);
+    if (startDelay > WARN_DELAY) {
```

Comment:
this is a minor comment, but in the event that there are a lot of persists occurring, we'll get spammed with these messages. But I think when so many persists are occurring, we'll have other problems